tools/harness-simulation/posix/sniffer_sim/sniffer.py

```diff
@@ -28,104 +28,172 @@
-        # Create transport
-        transport_factory = sniffer_transport.SnifferTransportFactory()
-        self._transport = transport_factory.create_transport()
+    class State(enum.Enum):
+        STOPPED = 0
+        RUNNING = 1
 
+    def _clear(self):
+        self._state = SnifferServicer.State.STOPPED
+        self._pcap = None
+        self._allowed_nodeids = None
+        self._transport = None
         self._thread = None
-        self._thread_alive = threading.Event()
         self._thread_alive.clear()
 
+    def __init__(self):
+        self._thread_alive = threading.Event()
+        self._mutex = threading.Lock()  # for self._allowed_nodeids
+        self._clear()
+
     def _sniffer_main_loop(self):
         """ Sniffer main loop. """
 
         self.logger.debug('Sniffer started.')
 
         while self._thread_alive.is_set():
+            # Avoid being blocked endlessly when there is no data
+            if not self._transport.ready(0.1):
+                continue
             data, nodeid = self._transport.recv(self.RECV_BUFFER_SIZE)
-            self._pcap.append(data)
+
+            self._mutex.acquire()
+            allowed_nodeids = self._allowed_nodeids
+            self._mutex.release()
+
+            # Equivalent to RF enclosure
+            if nodeid in allowed_nodeids:
+                self._pcap.append(data)
 
         self.logger.debug('Sniffer stopped.')
 
-    def start(self):
+    def Start(self, request, context):
         """ Start sniffing. """
 
+        self.logger.debug('call Start')
+
+        # Validate and change the state
+        if self._state != SnifferServicer.State.STOPPED:
+            return sniffer_pb2.StatusReply(status=sniffer_pb2.OPERATION_ERROR)
+        self._state = SnifferServicer.State.RUNNING
+
+        self._pcap = pcap_codec.PcapCodec(request.channel)
+
+        # Sniffer all nodes in default, i.e. there is no RF enclosure
+        self._allowed_nodeids = set(range(1, self.MAX_NODES_NUM + 1))
+
+        # Create transport
+        transport_factory = sniffer_transport.SnifferTransportFactory()
+        self._transport = transport_factory.create_transport()
+
+        # Start the sniffer main loop thread
         self._thread = threading.Thread(target=self._sniffer_main_loop)
         self._thread.daemon = True
-
         self._transport.open()
-
         self._thread_alive.set()
         self._thread.start()
 
-    def stop(self):
-        """ Stop sniffing. """
+        return sniffer_pb2.StatusReply(status=sniffer_pb2.OK)
 
-        self._thread_alive.clear()
+    def FilterNodes(self, request_iterator, context):
+        """ Only sniffer the specified nodes. """
 
-        self._transport.close()
+        self.logger.debug('call FilterNodes')
+
+        # Validate the state
+        if self._state != SnifferServicer.State.RUNNING:
+            return sniffer_pb2.StatusReply(status=sniffer_pb2.OPERATION_ERROR)
+
+        allowed_nodeids = set([x.nodeid for x in request_iterator])
```

Comment:
```suggestion
        allowed_nodeids = set(x.nodeid for x in request_iterator)
```
I guess `[]` is not necessary

Reply:
It is not necessary. Removed the square brackets.